lib/Dialect/LLHD/Transforms/LowerProcesses.cpp

```diff
@@ -0,0 +1,279 @@
+    return false;
+  }
+
+  // Helper function to skip across empty blocks with only a single successor.
+  auto skipToMergePoint = [&](Block *block) -> std::pair<Block *, ValueRange> {
+    ValueRange operands;
+    while (auto branchOp = dyn_cast<cf::BranchOp>(block->getTerminator())) {
+      if (!block->without_terminator().empty())
+        break;
+      block = branchOp.getDest();
+      operands = branchOp.getDestOperands();
+      if (std::distance(block->pred_begin(), block->pred_end()) > 1)
+        break;
+      if (!operands.empty())
+        break;
+    }
+    return {block, operands};
+  };
+
+  // Ensure that the entry block and wait op converge on the same block and with
+  // the same block arguments.
+  auto &entry = processOp.getBody().front();
+  auto [entryMergeBlock, entryMergeArgs] = skipToMergePoint(&entry);
+  auto [waitMergeBlock, waitMergeArgs] = skipToMergePoint(waitOp.getDest());
+  if (entryMergeBlock != waitMergeBlock) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Skipping process " << processOp.getLoc()
+               << ": control from entry and wait does not converge\n");
+    return false;
+  }
+  if (entryMergeArgs != waitMergeArgs) {
+    LLVM_DEBUG(llvm::dbgs() << "Skipping process " << processOp.getLoc()
+                            << ": control from entry and wait converges with "
+                               "different block arguments\n");
+    return false;
+  }
+
+  // Ensure that no values are live across the wait op.
+  Liveness liveness(processOp);
+  for (auto value : liveness.getLiveOut(waitOp->getBlock())) {
+    if (value.getParentRegion()->isProperAncestor(&processOp.getBody()))
+      continue;
+    LLVM_DEBUG({
+      llvm::dbgs() << "Skipping process " << processOp.getLoc() << ": value ";
+      value.print(llvm::dbgs(), OpPrintingFlags().skipRegions());
+      llvm::dbgs() << " live across wait\n";
+    });
+    return false;
+  }
+
+  return true;
+}
+
+/// Mark values the process observes that are defined outside the process.
+void Lowering::markObservedValues() {
+  SmallVector<Value> worklist;
+  auto markObserved = [&](Value value) {
+    if (observedValues.insert(value).second)
+      worklist.push_back(value);
+  };
+
+  for (auto value : waitOp.getObserved())
+    if (value.getParentRegion()->isProperAncestor(&processOp.getBody()))
+      markObserved(value);
+
+  while (!worklist.empty()) {
+    auto value = worklist.pop_back_val();
+    auto *op = value.getDefiningOp();
+    if (!op)
+      continue;
+
+    // Look through probe ops to mark the probe signal as well, just in case
+    // there may be multiple probes of the same signal.
+    if (auto probeOp = dyn_cast<PrbOp>(op))
+      markObserved(probeOp.getSignal());
+
+    // Look through operations that simply reshape incoming values into an
+    // aggregate form from which any changes remain apparent.
+    if (isa<hw::ArrayCreateOp, hw::StructCreateOp, comb::ConcatOp>(op))
```

Comment:
`hw.bitcast` is another one I have seen a few times in the past. Mostly with a constant as input, however, so that case should already be covered I think.

Reply:
Oh yeah good thinking, let me add that!